### PR TITLE
Tapp 203

### PR DESCRIPTION
**Pull request by thezane**
Split up new position form into component for defining page structure and view for accessing redux store.  https://github.com/uoft-tapp/tapp/issues/203

### CODE REVIEW

File: frontend/src/views/new_position_form/reducer.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { createReducer } from "redux-create-reducer";
+import {
+    CREATE_NEW_POSITION_SUCCESS,
+    IMPORT_NEW_POSITION_FAILURE,
+    IMPORT_NEW_POSITION_SUCCESS
+} from "../../api/constants/new_position";
+
+const initialState = {
+    list: [],
```
Comment (siefkenj) [inline frontend/src/views/new_position_form/reducer.js]: `list` is not a good name...what is it a list of?

File: frontend/src/components/new-position-form.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
+import React from "react";
+import { moment } from "moment";
+
+const toNames = ({ first_name, last_name }) => `${first_name} ${last_name}`;
+
+const validNumber = val => (val && isNaN(val) ? "Not a valid number" : false);
+const validDate = val =>
+    (val && moment(val, "YYYY-MM-DD", true).isValid()) || val === ""
+        ? false
+        : "Not a valid date";
```
Comment (siefkenj) [inline frontend/src/components/new-position-form.js]: What?? `validDate === false` means that the date is valid? Don't mix variables that store booleans and strings. If it is supposed to be an error message, make it an error message.

File: frontend/src/components/new-position-form.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+// check for format XXX123H1
+const validCourseCode = val => {
```
Comment (siefkenj) [inline frontend/src/components/new-position-form.js]: No. Don't do this check. The backend will say whether a course code is valid. If it exists in the database, it's valid.

File: frontend/src/components/new-position-form.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        !isNaN(val.slice(3, 6)) &&
+        (val[6] === "H" || val[6] === "Y") &&
+        !isNaN(val[7])
+    ) {
+        return false;
+    } else {
+        return "Not a valid course code";
+    }
+};
+const validInstructor = (val, props) => {
```
Comment (siefkenj) [inline frontend/src/components/new-position-form.js]: You're creating needless datastructures...This function should also be called `existingInstructor`. Any of these functions that might be used more than once should be in the `utils` directory.

File: frontend/src/components/new-position-form.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    const instructors = props.instructors.map(toNames);
+    return instructors.indexOf(val) !== -1 ? false : "Not a valid instructor";
+};
+
+export const newPositionFields = [
+    {
+        label: "Course Code",
+        value: "course_code",
+        validate: [validCourseCode],
```
Comment (siefkenj) [inline frontend/src/components/new-position-form.js]: Why are the validators an array? Do we need to chain validators?

File: frontend/src/components/new-position-form.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        <label style={{ width: "100px" }}>
+            {required ? label + "*" : label}
+        </label>
+        <input value={curValue} onChange={onChange} required={required} />
+    </div>
+);
+
+export const TextboxInput = ({ label, curValue, onChange, required }) => (
+    <div>
+        <label style={{ width: "100px" }}>
```
Comment (siefkenj) [inline frontend/src/components/new-position-form.js:117]: Don't do inline styles.

File: frontend/src/components/new-position-form.js [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    </div>
+);
+
+export const InstructorList = ({ curValue, instructors, setInstructor }) => {
+    const valid = validInstructor(curValue, { instructors });
+    const matchingInstructors = instructors
+        .map(({ first_name, last_name }) => `${first_name} ${last_name}`)
+        .filter(
+            name => name.toLowerCase().indexOf(curValue.toLowerCase()) !== -1
+        );
```
Comment (siefkenj) [inline frontend/src/components/new-position-form.js]: This confuses me...how could there me more than one matching instructor?

File: frontend/src/api/actions/new_position.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    IMPORT_NEW_POSITION_FAILURE,
+    IMPORT_NEW_POSITION_SUCCESS
+} from "../constants/new_position";
+import { apiPOST } from "../../libs/apiUtils";
+
+export const createNewPosition = payload => async dispatch => {
+    try {
+        const data = await apiPOST("/positions", payload);
+        dispatch(createNewPositionSuccess(data));
+        window.location = "/tapp/positions";
```
Comment (siefkenj) [inline frontend/src/api/actions/new_position.js]: Never, never, never set `window.location` yourself. API actions should be side-effect free. That is, they can change the model state, but nothing else (no changing of routes, no changing of UI state)

Comment (siefkenj) [reply frontend/src/api/actions/new_position.js]: When you dispatch something in the UI, you can `await` the dispatch. That is the appropriate place to change the route.

**Comment by siefkenj** [comment]
Put the details in this PR. It's no fun to have to click around links to see what is supposed to be happening...

File: frontend/src/components/new-position-form.js (hunk 6 above)
Comment (thezane) [reply frontend/src/components/new-position-form.js:117]: The styling for this page will be changed in a different issue.  When that happens, inline styles will be removed.

**Comment by siefkenj** [comment]
Merge conflict now...

File: frontend/src/views/new_position_form/reducer.js (hunk 1 above)
Comment (thezane) [reply frontend/src/views/new_position_form/reducer.js]: Renamed

File: frontend/src/components/new-position-form.js (hunk 2 above)
Comment (thezane) [reply frontend/src/components/new-position-form.js]: A validator now returns true iff a field is valid.  If a field is invalid, an err msg is accumulated in `fillablePosition.jsx`.

File: frontend/src/components/new-position-form.js (hunk 3 above)
Comment (thezane) [reply frontend/src/components/new-position-form.js]: Removed check

File: frontend/src/components/new-position-form.js (hunk 5 above)
Comment (thezane) [reply frontend/src/components/new-position-form.js]: Replaced array of validators with single validator

File: frontend/src/components/new-position-form.js (hunk 4 above)
Comment (thezane) [reply frontend/src/components/new-position-form.js]: Removed constructed list via map.  Renamed function.

File: frontend/src/api/actions/new_position.js (hunk 8 above)
Comment (thezane) [reply frontend/src/api/actions/new_position.js]: Moved route changing to view

File: frontend/src/components/new-position-form.js (hunk 7 above)
Comment (thezane) [reply frontend/src/components/new-position-form.js]: The map & filter code is part of an autocompletion feature that lists instructors whose name is a superstring of what the user typed.  So yes, more than 1 instructor can survive the filter.  I rewrote it so we only have 1 loop.

Screenshot from 2019-10-10 21-28-19

**Comment by siefkenj** [comment]
Looks good! I'll merge and any remaining issues can be fixed later.

**Pull request merged**